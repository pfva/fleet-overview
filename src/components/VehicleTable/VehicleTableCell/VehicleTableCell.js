import { useState, useRef, useEffect, useContext } from 'react';
import VehicleContext from '../../../../src/store/vehicle-context';
import VehicleStatus from './VehicleStatus/VehicleStatus';
import VehicleEquipments from './VehicleEquipments/VehicleEquipments';
import styles from './VehicleTableCell.module.css';

const VehicleTableCell = ({ content: { type, data, vehicleId } }) => {
  const [updatedData, setUpdatedData] = useState(data);
  const [isEditing, setIsEditing] = useState(false);
  const inputElement = useRef(null);
  const vehicleContext = useContext(VehicleContext);

  const toggleEditHandler = () => {
    setIsEditing(!isEditing);
  };

  const inputChangeHandler = e => {
    setUpdatedData(e.target.value);
  };

  // Closes "editing mode" if user clicks anywhere outside the input field
  useEffect(() => {
    const handleClickOutsideInput = event => {
      if (
        inputElement.current &&
        !inputElement.current.contains(event.target)
      ) {
        setIsEditing(false);
        setUpdatedData(data);
      }
    };

    document.addEventListener('mousedown', handleClickOutsideInput);
    return () => {
      document.removeEventListener('mousedown', handleClickOutsideInput);
    };
  }, [inputElement, data]);

  // Updates the context state if the user has changed the data
  useEffect(() => {
    const handleEnterKeypress = e => {
      if (e.code === 'Enter') {
        setIsEditing(false);
        if (updatedData !== data) {
          const vehicleToUpdate = vehicleContext.vehicles.find(
            vehicle => vehicle.id === vehicleId
          );
          const updatedVehicle = {
            ...vehicleToUpdate,
            [type]: updatedData,
          };
          vehicleContext.updateVehicle(vehicleId, updatedVehicle);
        }
      }
    };

    const handleEscKeypress = e => {
      if (e.code === 'Escape') {
        setIsEditing(false);
        setUpdatedData(data);
      }
    };

    document.addEventListener('keydown', handleEnterKeypress);
    document.addEventListener('keydown', handleEscKeypress);
    return () => {
      document.removeEventListener('keydown', handleEnterKeypress);
      document.removeEventListener('keydown', handleEscKeypress);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [updatedData]);

  // Special interactivity for equipments
  if (type === 'equipments' && Array.isArray(data)) {
    return <VehicleEquipments equipments={data} vehicleId={vehicleId} />;
  }

  // Special interactivity for status
  if (type === 'status') {
    return <VehicleStatus status={data} vehicleId={vehicleId} />;
  }

  return isEditing ? (
    <input
      type="text"
      value={updatedData}
      onChange={inputChangeHandler}
      ref={inputElement}
      className={styles.input}
      autoFocus
      data-testid="VehicleTableCell-input"
    />
  ) : (
    <p className={styles.vehicleTableCell} onClick={toggleEditHandler} data-testid="VehicleTableCell-text">
      <span className={styles.dataText}>{data}</span>
    </p>
  );
};

export default VehicleTableCell;
