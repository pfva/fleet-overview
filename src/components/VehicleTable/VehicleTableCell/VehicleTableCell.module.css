.vehicleTableCell {
  flex: 1 1 calc(100% / 6);
  margin: 0 var(--spacing-extra-small);
  text-align: left;
  cursor: pointer;
  user-select: none;
}

.vehicleEquipments {
  flex: 1 1 calc(100% / 6);
  display: flex;
  flex-direction: column;
}

.iconWrapper {
  display: flex;
}

.input {
  max-width: calc((100% / 6) - var(--spacing-small));
  max-height: var(--input-height);
  margin-right: var(--spacing-small);
  margin-bottom: -4px;
  margin-top: -4px;
}
