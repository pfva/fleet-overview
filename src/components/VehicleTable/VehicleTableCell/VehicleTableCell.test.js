import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import VehicleTableCell from './VehicleTableCell';

describe('VehicleTableCell', () => {
  it('renders without crashing', () => {
    const mockedId = { type: 'id', data: 'GA33' };
    const tree = renderer
      .create(<VehicleTableCell content={mockedId} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the name value', () => {
    const mockedName = { type: 'name', data: 'Patrick Star' };
    render(<VehicleTableCell content={mockedName} />);
    const name = screen.getByText(/Patrick/i);
    expect(name).toBeInTheDocument();
  });

  it('renders the status value and icon if status is active', () => {
    const mockedStatus = { type: 'status', data: 'active' };
    render(<VehicleTableCell content={mockedStatus} />);
    const status = screen.getByText(/active/i);
    const icon = screen.getByTestId(/Icon-active/i);
    expect(status).toBeInTheDocument();
    expect(icon).toBeInTheDocument();
  });

  it('renders the status value but not the icon if status is inactive', () => {
    const mockedStatus = { type: 'status', data: 'inactive' };
    render(<VehicleTableCell content={mockedStatus} />);
    const status = screen.getByText(/inactive/i);
    const icon = screen.queryByTestId(/Icon-active/i);
    expect(status).toBeInTheDocument();
    expect(icon).not.toBeInTheDocument();
  });

  it('renders no equipment name if equipments data is undefined', () => {
    const mockedEquipments = { type: 'equipments', data: undefined };
    const tree = renderer
      .create(<VehicleTableCell content={mockedEquipments} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the correct equipment name if the equipments array contains one value', () => {
    const mockedEquipments = { type: 'equipments', data: [2] };
    render(<VehicleTableCell content={mockedEquipments} />);
    const equipment = screen.getByText(/Tachograph/i);
    expect(equipment).toBeInTheDocument();
  });

  it('renders the correct equipment names if the equipments array contains multiple values', () => {
    const mockedEquipments = { type: 'equipments', data: [2, 3] };
    render(<VehicleTableCell content={mockedEquipments} />);
    const equipmentOne = screen.getByText(/Tachograph/i);
    const equipmentTwo = screen.getByText(/Fire Extinguisher/i);
    expect(equipmentOne).toBeInTheDocument();
    expect(equipmentTwo).toBeInTheDocument();
  });
});
