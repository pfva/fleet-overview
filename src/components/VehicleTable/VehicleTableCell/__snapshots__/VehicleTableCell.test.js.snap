// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleTableCell renders no equipment name if equipments data is undefined 1`] = `
<p
  className="vehicleTableCell"
  data-testid="VehicleTableCell-text"
  onClick={[Function]}
>
  <span
    className="dataText"
  />
</p>
`;

exports[`VehicleTableCell renders without crashing 1`] = `
<p
  className="vehicleTableCell"
  data-testid="VehicleTableCell-text"
  onClick={[Function]}
>
  <span
    className="dataText"
  >
    GA33
  </span>
</p>
`;
